pepdbagent/search.py

```diff
@@ -0,0 +1,240 @@
+        :param admin: True, if user is admin for this namespace
+        :return: number of found project in specified namespace
+        """
+        if not admin:
+            admin_str = f"""and ({PRIVATE_COL} = 'false')"""
+        else:
+            admin_str = ""
+        count_sql = f"""
+        select count(*)
+            from {DB_TABLE_NAME} 
+                where ({NAME_COL} ILIKE '%%{search_str}%%' or ({PROJ_COL}->>'description') ILIKE '%%{search_str}%%') 
+                    and {NAMESPACE_COL} = '{namespace}' {admin_str};
+        """
+        result = self._run_sql_fetchall(count_sql)
+        try:
+            number_of_prj = result[0][0]
+        except IndexError:
+            number_of_prj = 0
+        return number_of_prj
+
+    def _find_project(
+        self,
+        namespace: str,
+        search_str: str = "",
+        admin: bool = False,
+        limit: int = DEFAULT_LIMIT,
+        offset: int = DEFAULT_OFFSET,
+    ) -> List[ProjectSearchResultModel]:
+        """
+        Search for project inside namespace by providing search string.
+        :param namespace: namespace where to search for a project
+        :param search_str: search string that has to be found in the name or project description
+        :param admin: True, if user is admin of the namespace [Default: False]
+        :param limit: limit of return results
+        :param offset: number of results off set (that were already showed)
+        :return: list of found projects with their annotations.
+        """
+        if not admin:
+            admin_str = f"""and ({PRIVATE_COL} = 'false')"""
+        else:
+            admin_str = ""
+        count_sql = f"""
+        select {NAMESPACE_COL}, {NAME_COL}, {TAG_COL}, {N_SAMPLES_COL}, ({PROJ_COL}->>'description'), {DIGEST_COL}, {PRIVATE_COL}, {SUBMISSION_DATE_COL}, {LAST_UPDATE_DATE_COL}
+            from {DB_TABLE_NAME} 
+                where ({NAME_COL} ILIKE '%%{search_str}%%' or ({PROJ_COL}->>'description') ILIKE '%%{search_str}%%') 
+                    and {NAMESPACE_COL} = '{namespace}' {admin_str} 
+                        LIMIT {limit} OFFSET {offset};
+        """
+        results = self._run_sql_fetchall(count_sql)
+        results_list = []
+        try:
+            for res in results:
+                results_list.append(
+                    ProjectSearchResultModel(
+                        namespace=res[0],
+                        name=res[1],
+                        tag=res[2],
+                        number_of_samples=res[3],
+                        description=res[4],
+                        digest=res[5],
+                        is_private=res[6],
+                        last_update_date=str(res[8]),
+                        submission_date=str(res[7]),
+                    )
+                )
+        except KeyError:
+            results_list = []
+
+        return results_list
+
+    def _get_count_of_found_namespaces(
```

Comment:
why is the get count of found namespaces a different function from the get namespaces function?

Reply:
I would suggest: 

Write a single function that runs 2 queries: 1. retrieve the objects of interest; 2. counts them. Put these into a standardized output structure, and return that.

Reply:
> I don't understand why you made Connection and Search two different subclasses of BaseConnection.

This was my recommendation. I noticed he was duplicating code for the `Search` and `Connection` class, so I recommended make a base class that holds the `_run_sql_fetchall` type code.